examples/counter/crate/src/lib.rs

```diff
@@ -28,59 +28,53 @@ impl Render for Counter {
     {
         let val = bumpalo::format!(in bump, "{}", self.val);
 
-        let increment = Node::element(
-            bump,
-            "button",
-            [Attribute {
-                name: "data-action",
-                value: "increment",
-            }],
-            [Node::text("+")],
-        );
-
-        let decrement = Node::element(
-            bump,
-            "button",
-            [Attribute {
-                name: "data-action",
-                value: "decrement",
-            }],
-            [Node::text("-")],
-        );
-
         Node::element(
             bump,
             "div",
             [],
-            [increment, Node::text(val.into_bump_str()), decrement],
+            [],
+            [
+                Node::element(
+                    bump,
+                    "button",
+                    [on(bump, "click", |root, vdom, _event| {
+                        root.unwrap_mut::<Counter>().increment();
+                        vdom.schedule_render();
+                    })],
+                    [],
+                    [Node::text("+")],
+                ),
+                Node::text(val.into_bump_str()),
+                Node::element(
+                    bump,
+                    "button",
+                    [on(bump, "click", |root, vdom, _event| {
+                        root.unwrap_mut::<Counter>().decrement();
```

Comment:
So just to confirm again from before, how would "middlewary style" libraries end up handling this? They don't know the type of the root object (right?), or would they have their own root vdom objects they can manipulate?

Reply:
They would be parameterized by some trait and would translate raw web events into "semantic" events for that component, and these semantic events would have a corresponding trait method.

I have TodoMVC implemented in the next PR after this, and it does this with its single `Todo` component, which you can sneak a peak at here: https://github.com/fitzgen/dodrio/blob/fill-out-examples/examples/todomvc/src/todo.rs#L9-L42